app/src/server/api/routers/billing.router.ts

```diff
@@ -0,0 +1,147 @@
+import { sql } from "kysely";
+import { z } from "zod";
+import { createTRPCRouter, protectedProcedure } from "~/server/api/trpc";
+import { kysely, prisma } from "~/server/db";
+import { v4 as uuidv4 } from "uuid";
+import { requireCanViewProject } from "~/utils/accessControl";
+import { toUTC } from "~/utils/dayjs";
+import { success } from "~/utils/errorHandling/standardResponses";
+import { getStats } from "./usage.router";
+
+export const billingRouter = createTRPCRouter({
+  invoices: protectedProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      await requireCanViewProject(input.projectId, ctx);
+
+      const invoices = await kysely
+        .selectFrom("Invoice as i")
+        .where("i.projectId", "=", input.projectId)
+        .orderBy("i.createdAt", "desc")
+        .selectAll()
+        .execute();
+
+      return { invoices };
+    }),
+  invoice: protectedProcedure
+    .input(
+      z.object({
+        invoiceId: z.string(),
+      }),
+    )
+    .query(async ({ input, ctx }) => {
+      const invoice = await prisma.invoice.findUniqueOrThrow({
+        where: { id: input.invoiceId },
+      });
+
+      await requireCanViewProject(invoice.projectId, ctx);
+
+      return { invoice };
+    }),
+  // This endpoint is for testing purposes only. It should be refactored and executed as a cron job.
+  createInvoice: protectedProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+      }),
+    )
+    .mutation(async ({ input, ctx }) => {
+      await requireCanViewProject(input.projectId, ctx);
+
+      const [startOfLastMonth, endOfPreviousMonth] = getLastMonthPeriod();
+
+      await kysely.transaction().execute(async (tx) => {
+        const baseQuery = tx
```

Comment:
For consistency I would flip this around: first I would update the `UsageLog` table to associate the correct items with the invoice, and then I would query explicitly the `UsageLog` entries that are associated with the invoice to get the stats.

Theoretically that should have the same behavior as what you have here, but I'd be much more comfortable knowing for sure that the usage logs we're using to compute the total are definitely the exact same ones we're associating with the invoice.

Reply:
It adds another step in this transaction, but it is a valid point. Changed it.